dask/array/core.py

```diff
@@ -880,31 +884,85 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
         raise ValueError("Different number of sources [%d] and targets [%d] than regions [%d]"
                          % (len(sources), len(targets), len(regions)))
 
-    updates = {}
-    keys = []
+    # Optimize all sources together
+    sources_dsk = sharedict.merge(*[e.__dask_graph__() for e in sources])
+    sources_dsk = Array.__dask_optimize__(
+        sources_dsk,
+        [e.__dask_keys__() for e in sources]
+    )
+
+    tgt_dsks = []
+    store_keys = []
+    store_dsks = []
+    if keep:
+        load_names = []
+        load_dsks = []
     for tgt, src, reg in zip(targets, sources, regions):
         # if out is a delayed object update dictionary accordingly
         try:
-            dsk = {}
-            dsk.update(tgt.dask)
+            each_tgt_dsk = {}
+            each_tgt_dsk.update(tgt.dask)
             tgt = tgt.key
         except AttributeError:
-            dsk = {}
+            each_tgt_dsk = {}
 
-        update = insert_to_ooc(src, tgt, lock=lock, region=reg)
-        keys.extend(update)
+        src = Array(
+            sources_dsk,
+            src.name,
+            src.chunks,
+            src.dtype
+        )
```

Comment:
Just a stylistic choice, I personally prefer to do all this one line. I find the ability to read isn't harmed, and the reduced number of lines makes the overall function easier to grasp. There are a few places in this PR where I'd change this. No strong thoughts though, up to you.

```python
        src = Array(sources_dsk, src.name, src.chunks, src.dtype)
```

Reply:
Yeah think I ran into issues with line length/`flake8` at some point. Can give it another look though as the code has changed over time.

Reply:
In cases like those I usually prefer something like

```python
        src = Array(sources_dsk, src.name,
                    src.chunks, src.dtype)
```

to minimize number of lines. But again, all personal preference.

Reply:
Seems this now fits on one line. Probably was spread out due to older less clear code. Now on one line.